File: Day09/C++/list_interview.cpp
#include <iostream>
using namespace std;

class SinglyList {
private:
    class Node {
    public:
        int data;
        Node* next;
        Node(int val) : data(val), next(nullptr) {}
    };

    Node* head;

public:
    SinglyList() : head(nullptr) {}

    void display() {
        cout << "List: ";
        Node* trav = head;
        while (trav != nullptr) {
            cout << trav->data << " -> ";
            trav = trav->next;
        }
        cout << endl;
    }

    void addLast(int val) {
        Node* newNode = new Node(val);
        if (head == nullptr) {
            head = newNode;
        } else {
            Node* trav = head;
            while (trav->next != nullptr) {
                trav = trav->next;
            }
            trav->next = newNode;
        }
    }

    void selectionSort() {
        for (Node* i = head; i != nullptr; i = i->next) {
            for (Node* j = i->next; j != nullptr; j = j->next) {
                if (i->data > j->data) {
                    int temp = i->data;
                    i->data = j->data;
                    j->data = temp;
                }
            }
        }
    }

    void revDisplay(Node* trav) {
        if (trav == nullptr) {
            return;
        }
        revDisplay(trav->next);
        cout << trav->data << ", ";
    }

    void revDisplay() {
        cout << "Rev : ";
        revDisplay(head);
        cout << endl;
    }

    ~SinglyList() {
        Node* trav = head;
        while (trav != nullptr) {
            Node* temp = trav;
            trav = trav->next;
            delete temp;
        }
    }
};

int main() {
    SinglyList list;
    list.addLast(30);
    list.addLast(50);
    list.addLast(10);
    list.addLast(40);
    list.addLast(20);
    list.display(); // 30 -> 50 -> 10 -> 40 -> 20 -> 
    list.selectionSort();
    list.display(); // 10 -> 20 -> 30 -> 40 -> 50 -> 
    list.revDisplay();
    return 0;
}

File: Day09/Python/list_interview.py
class SinglyList:
    class Node:
        def __init__(self, val):
            self.data = val
            self.next = None

    def __init__(self):
        self.head = None

    def display(self):
        print("List: ", end="")
        trav = self.head
        while trav is not None:
            print(trav.data, "-> ", end="")
            trav = trav.next
        print()

    def addLast(self, val):
        newNode = self.Node(val)
        if self.head is None:
            self.head = newNode
        else:
            trav = self.head
            while trav.next is not None:
                trav = trav.next
            trav.next = newNode

    def selectionSort(self):
        i = self.head
        while i is not None:
            j = i.next
            while j is not None:
                if i.data > j.data:
                    i.data, j.data = j.data, i.data
                j = j.next
            i = i.next

    def revDisplay(self, trav=None):
        if trav is None:
            print("Rev : ", end="")
            self.revDisplay(self.head)
            print()
            return
        if trav is None:
            return
        self.revDisplay(trav.next)
        print(trav.data, end=", ")


def main():
    list = SinglyList()
    list.addLast(30)
    list.addLast(50)
    list.addLast(10)
    list.addLast(40)
    list.addLast(20)
    list.display()  # 30 -> 50 -> 10 -> 40 -> 20 -> 
    list.selectionSort()
    list.display()  # 10 -> 20 -> 30 -> 40 -> 50 -> 
    list.revDisplay()


if __name__ == "__main__":
    main()